### PR TITLE
Eliminate Fallthrough typeclass

### PR DESCRIPTION
**Pull request by rossabaker**
Introduces a MaybeResponse, which has a monoid similar to the "first
option" monoid.  This lets us compose services according to the standard
Kleisli monoid instead of our ad hoc concept.

Furthermore, it forces us to carefully consider the effect of
fallthroughs in middleware, which has already bitten us once (#803).

### CODE REVIEW

File: core/src/main/scala/org/http4s/Message.scala [hunk 1]
```diff
@@ -224,6 +225,53 @@ object Request {
   }
 }
 
+/**
+ * Represents that a service either returns a [[Response]] or a [[Pass]] to fall through
+ * to another service.
+ */
+sealed trait MaybeResponse {
```
Comment (bryce-anderson) [inline core/src/main/scala/org/http4s/Message.scala]: Looks like there are a number of places that would be well served by a `notReallyAFlatMapButFeelsKindOfLikeOne(f: Response => MaybeResponse)` method. Does that form of method have a common name?

Comment (rossabaker) [inline core/src/main/scala/org/http4s/Message.scala]: Not that I know of.

**Comment by bryce-anderson** [comment]
There are a lot of changes in here so perhaps I've missed something, but in general, I like this.

**Comment by rossabaker** [comment]
The cats branch's server project is failing on a fallthrough test case.  I'm hopeful it gets cleaned up by this.

**Comment by rossabaker** [comment]
Build failure is #817.

**Pull request merged**